File: apps/web/lib/services/commentService.ts
import { apiFetch } from "./apiClient";
import type {
  CommentItem,
  CommentListResponse,
  CreateCommentRequest,
  UserCommentActivityPage,
} from "@/lib/types/comment";

export const commentService = {
  list: (postId: string) =>
    apiFetch<CommentListResponse>(`/api/v1/posts/${postId}/comments`),

  listByUser: (userId: string, page = 1, size = 20) =>
    apiFetch<UserCommentActivityPage>(
      `/api/v1/users/${encodeURIComponent(userId)}/comments?page=${page}&size=${size}`,
    ),

  create: (postId: string, body: CreateCommentRequest) =>
    apiFetch<CommentItem>(`/api/v1/posts/${postId}/comments`, {
      method: "POST",
      body,
    }),
};

File: apps/server/src/main/java/com/chtholly/comment/service/CommentService.java
package com.chtholly.comment.service;

import com.chtholly.common.api.pagination.PageResponse;
import com.chtholly.comment.api.dto.CommentResponse;
import com.chtholly.comment.api.dto.CreateCommentRequest;
import com.chtholly.comment.api.dto.UserCommentActivityResponse;

import java.util.List;
import java.util.Map;

public interface CommentService {

    PageResponse<CommentResponse> listByPost(long postId, Long viewerUserIdNullable, int page, int size);

    /**
     * Lists one user's comments that are visible on published public posts.
     *
     * @param userId comment author ID
     * @param page page number (1-based)
     * @param size page size
     * @return paginated public comment activity
     */
    PageResponse<UserCommentActivityResponse> listByUser(long userId, int page, int size);

    /**
     * Counts active comment rows for each post, including replies.
     *
     * @param postIds post IDs to count
     * @return one non-negative count for every distinct non-null post ID
     */
    Map<Long, Long> countActiveByPostIds(List<Long> postIds);

    CommentResponse create(long postId, long userId, CreateCommentRequest request);

    void delete(long postId, long commentId, long userId);

    void adminDelete(long postId, long commentId);
}
